chain/network/src/types.rs

```diff
@@ -104,33 +109,56 @@ impl PeerInfo {
     }
 }
 
+// Note, `Display` automatically implements `ToString` which must be reciprocal to `FromStr`.
 impl fmt::Display for PeerInfo {
     fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
-        if let Some(acc) = self.account_id.as_ref() {
-            write!(f, "({}, {:?}, {})", self.id, self.addr, acc)
+        write!(f, "{}", self.id)?;
+        if let Some(addr) = &self.addr {
+            write!(f, "@{}", addr)?;
+        }
+        if let Some(account_id) = &self.account_id {
+            write!(f, "@{}", account_id)?;
+        }
+        Ok(())
+    }
+}
+
+impl FromStr for PeerInfo {
+    type Err = Box<dyn std::error::Error>;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        let chunks: Vec<&str> = s.split('@').collect();
+        let addr;
+        let account_id;
+        if chunks.len() == 1 {
+            addr = None;
```

Comment:
hm, why if there is no @ it's id not an addr?

Reply:
The previous `TryFrom<&str>` implementation returns an error if there is anything but `id@ip_addr`. I have relaxed it by allowing `id`, `id@ip_addr`, `id@account_id`, `id@ip_addr@account_id`. AFAIR we can transitively have `PeerInfo` without `addr` but with `id`, I will let @mfornet to correct me here. So I concluded that this relaxation makes sense.